File: C++/261. Graph Valid Tree.cpp
//261. Graph Valid Tree

class Solution {
public:
    bool validTree(int n, vector<pair<int, int>>& edges) {
        vector<int> nodes(n,0);
        for(int i=0;i<nodes.size();++i){
            nodes[i] = i;
        }
        for(int i=0;i<edges.size();++i){
            int root1 = edges[i].first;
            int root2 = edges[i].second;
            while(nodes[root1]!=root1)
                root1 = nodes[root1];
            while(nodes[root2]!=root2)
                root2 = nodes[root2];
            if(root1==root2)
                return false;
            nodes[root2] = root1;
        }
        return  edges.size() == n-1;
    }
};

File: python/261. Graph Valid Tree.py
class Solution:
    def validTree(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: bool
        """
        if len(edges) != n-1:
            return False
        graph = collections.defaultdict(list)
        for x,y in edges:
            graph[x] += y,
            graph[y] += x,
        
        q = collections.deque([0])
        visited = set()
        while q:
            node = q.pop()
            if node in visited:
                return False
            visited.add(node)
            for nbr in graph[node]:
                q.append(nbr)
                graph[nbr].pop(graph[nbr].index(node))
                
        return all([i in visited for i in range(n)])